File: Others/binary_search/time-based-key-value-store.js
// 981. Time Based Key-Value Store
// https://leetcode.com/problems/time-based-key-value-store/description/

var TimeMap = function () {
  this.hashMap = new Map();
};

TimeMap.prototype.set = function (key, value, timestamp) {
  if (this.hashMap.has(key)) {
    let arr = this.hashMap.get(key);
    arr.push([timestamp, value]);
    this.hashMap.set(key, arr);
  } else {
    let arr = [timestamp, value];
    this.hashMap.set(key, [arr]);
  }
  return null;
};

TimeMap.prototype.get = function (key, timestamp) {
  // Base case 1 & 2: key does not exist, or no prev timesetamps, return ""
  let arr = this.map.get(key);
  if (!this.map.has(key) || arr[0][0] > timestamp) return "";

  // Use Binary Search to look for the required value
  const index = binarySearch(arr, timestamp);
  if (arr[index][0] > timestamp) return arr[index - 1][1];
  return arr[index][1];
};

function binarySearch(arr, timestamp) {
  let start = 0,
    end = arr.length - 1,
    mid;
  // Iterate while start not meets end
  while (start <= end) {
    // Find the mid index
    mid = Math.floor((start + end) / 2);

    // If element is present at mid, return True
    if (arr[mid][0] === timestamp) return mid;
    // Else look in right
    else if (arr[mid][0] < timestamp) start = mid + 1;
    // Or look left
    else end = mid - 1;
  }
  return mid;
}

File: binary_search/time-based-key-value-store.java
// https://leetcode.com/problems/time-based-key-value-store/description/

// Java
class TimeMap {

    public Map<String, ArrayList<Time>> map;

    public TimeMap() {
        this.map = new HashMap<>();
    }
    
    public void set(String key, String value, int timestamp) {
        // key doesn't exist yet, add a new list
        if (!map.containsKey(key)) {
            ArrayList<Time> newList = new ArrayList<>();
            newList.add(new Time(value, timestamp));
            map.put(key, newList);
        } else {
            ArrayList<Time> currList = map.get(key);

            // Binary search to find insertion point
            int l = 0, r = currList.size();

            while (l < r) {
                int mid = (r + l) / 2;
                if (timestamp > currList.get(mid).timestamp) l = mid + 1;
                else r = mid;
            }

            // If timestamp equals existing one, overwrite
            if (l < currList.size() && currList.get(l).timestamp == timestamp) {
                currList.set(l, new Time(value, timestamp));
            } else {
                currList.add(l, new Time(value, timestamp));
            }
        }
    }
    
    public String get(String key, int timestamp) {
        if (!map.containsKey(key)) return "";

        ArrayList<Time> currList = map.get(key);

        int l = 0, r = currList.size() - 1;
        String res = "";

        while (l <= r) {
            int mid = (r + l) / 2;
            if (currList.get(mid).timestamp <= timestamp) {
                res = currList.get(mid).value;
                l = mid + 1;
            }
            else r = mid - 1; 
        }

        return res;
    }
}

class Time {
    public String value;
    public int timestamp;

    public Time (String value, int timestamp) {
        this.value = value;
        this.timestamp = timestamp;
    }
}